sdk/src/azure/iot/az_iot_pnp_client_twin.c

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <azure/iot/az_iot_pnp_client.h>
+
+#include <azure/core/internal/az_precondition_internal.h>
+#include <azure/core/internal/az_result_internal.h>
+
+static const az_span iot_hub_twin_desired = AZ_SPAN_LITERAL_FROM_STR("desired");
+static const az_span iot_hub_twin_desired_version = AZ_SPAN_LITERAL_FROM_STR("$version");
+static const az_span property_response_value_name = AZ_SPAN_LITERAL_FROM_STR("value");
+static const az_span property_ack_code_name = AZ_SPAN_LITERAL_FROM_STR("ac");
+static const az_span property_ack_version_name = AZ_SPAN_LITERAL_FROM_STR("av");
+static const az_span property_ack_description_name = AZ_SPAN_LITERAL_FROM_STR("ad");
+static const az_span component_property_label_name = AZ_SPAN_LITERAL_FROM_STR("__t");
+static const az_span component_property_label_value = AZ_SPAN_LITERAL_FROM_STR("c");
+
+AZ_NODISCARD az_result az_iot_pnp_client_twin_parse_received_topic(
+    az_iot_pnp_client const* client,
+    az_span received_topic,
+    az_iot_pnp_client_twin_response* out_response)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_VALID_SPAN(received_topic, 1, false);
+  _az_PRECONDITION_NOT_NULL(out_response);
+
+  az_iot_hub_client_twin_response hub_twin_response;
+  _az_RETURN_IF_FAILED(az_iot_hub_client_twin_parse_received_topic(
+      &client->_internal.iot_hub_client, received_topic, &hub_twin_response));
+
+  out_response->request_id = hub_twin_response.request_id;
+  out_response->response_type
+      = (az_iot_pnp_client_twin_response_type)hub_twin_response.response_type;
+  out_response->status = hub_twin_response.status;
+  out_response->version = hub_twin_response.version;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_pnp_client_twin_property_begin_component(
+    az_iot_pnp_client const* client,
+    az_json_writer* json_writer,
+    az_span component_name)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(json_writer);
```

Comment:
Would it be useful to add more preconditions on which state you expect the provided json writer to be in? For example, if we just wrote `[`, this will fail the `az_json_writer_append_property_name` precondition.

Or is it better to just rely on the writer's validation/preconditions?